packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -2418,11 +2418,40 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     updateKeepAlive();
   }
 
+  String _cachedText = '';
+  double _cachedScrollOffset = 0.0;
+  Rect? _cachedFirstRect;
+
+  void _updateSelectionRects() {
+    if (defaultTargetPlatform != TargetPlatform.iOS)
+      return;
+    final TextSpan textSpan = buildTextSpan();
+    final StringBuffer text = StringBuffer();
+    textSpan.computeToPlainText(text);
+    final Rect firstRect = renderEditable.getBoxesForSelection(const TextSelection(baseOffset: 0, extentOffset: 1)).first.toRect();
+    final double scrollOffset = _scrollController?.offset ?? 0.0;
+    final ScrollDirection scrollDirection = _scrollController?.position.userScrollDirection ?? ScrollDirection.idle;
+    if (scrollDirection == ScrollDirection.idle && text.toString() != _cachedText ||
+        scrollOffset != _cachedScrollOffset ||
+        _cachedFirstRect != firstRect) {
+      _cachedText = text.toString();
+      _cachedScrollOffset = scrollOffset;
+      _cachedFirstRect = firstRect;
+      final Offset rectOffset = Offset(_isMultiline ? 0.0 : -1 * scrollOffset, _isMultiline ? -1 * scrollOffset : 0.0);
+      final List<Rect> rects = List<Rect>.generate(
+              text.length, (int i) => renderEditable.getBoxesForSelection(TextSelection(baseOffset: i, extentOffset: i + 1)).first.toRect())
+          .map((Rect rect) => rect.translate(rectOffset.dx, rectOffset.dy))
```

Comment:
this should probably happen in the `RenderEditable` public api you exposed in the PR.

Reply:
So using `localToGlobal` on the render object there instead of offsetting by the scroll offset here?

Reply:
The `RenderEditable.getBoxesForSelection` API you exposed in the PR should be responsible for doing the `translate(rectOffset.dx, rectOffset.dy)` thing. The caller would usually expect the `TextBox`es to be described in the `RenderEditable`'s coordiante space instead of `_textPainter`'s.